### PR TITLE
Usage of C++ `<thread>` header requires linking to threading library

### PR DESCRIPTION
**Pull request by LebedevRI**
Usage of C++ `<thread>` header requires linking to threading library

`Generator.cpp` and `ThreadPool.h` both `#include <thread>`,
but don't link to the threading implementation.

This fixes build for me on debian sid, which is failing otherwise with:
```
$ ninja
[  0% 2/1480][  0% 0:00:00 + 0:33:06] Linking CXX executable src/autoschedulers/adams2019/get_host_target
FAILED: src/autoschedulers/adams2019/get_host_target
: && /usr/local/bin/clang++ -pipe -O3 -DNDEBUG  src/autoschedulers/adams2019/CMakeFiles/get_host_target.dir/get_host_target.cpp.o -o src/autoschedulers/adams2019/get_host_target  -Wl,-rpath,/repositories/halide/build/src:  src/libHalide.so.13.0.0 && :
ld: error: /usr/lib/gcc/x86_64-linux-gnu/11/libstdc++.so: undefined reference to pthread_create [--no-allow-shlib-undefined]
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[  0% 5/1480][  0% 0:00:00 + 0:12:47] Linking CXX executable src/autoschedulers/adams2019/test_apps_autoscheduler
FAILED: src/autoschedulers/adams2019/test_apps_autoscheduler
: && /usr/local/bin/clang++ -pipe -O3 -DNDEBUG  src/autoschedulers/adams2019/CMakeFiles/test_apps_autoscheduler.dir/test.cpp.o -o src/autoschedulers/adams2019/test_apps_autoscheduler  -Wl,-rpath,/repositories/halide/build/src  src/libHalide.so.13.0.0  -ldl && :
ld: error: /usr/lib/gcc/x86_64-linux-gnu/11/libstdc++.so: undefined reference to pthread_create [--no-allow-shlib-undefined]
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[  0% 7/1480][  0% 0:00:00 + 0:08:53] Linking CXX executable src/autoschedulers/adams2019/test_function_dag
FAILED: src/autoschedulers/adams2019/test_function_dag
: && /usr/local/bin/clang++ -pipe -O3 -DNDEBUG  src/autoschedulers/adams2019/CMakeFiles/test_function_dag.dir/test_function_dag.cpp.o src/autoschedulers/adams2019/CMakeFiles/test_function_dag.dir/FunctionDAG.cpp.o src/autoschedulers/adams2019/CMakeFiles/test_function_dag.dir/ASLog.cpp.o -o src/autoschedulers/adams2019/test_function_dag  -Wl,-rpath,/repositories/halide/build/src  src/libHalide.so.13.0.0 && :
ld: error: /usr/lib/gcc/x86_64-linux-gnu/11/libstdc++.so: undefined reference to pthread_create [--no-allow-shlib-undefined]
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[  0% 9/1480][  0% 0:00:00 + 0:06:41] Linking CXX executable src/autoschedulers/li2018/gradient_autoscheduler_test_cpp
FAILED: src/autoschedulers/li2018/gradient_autoscheduler_test_cpp
: && /usr/local/bin/clang++ -pipe -O3 -DNDEBUG  src/autoschedulers/li2018/CMakeFiles/gradient_autoscheduler_test_cpp.dir/test.cpp.o -o src/autoschedulers/li2018/gradient_autoscheduler_test_cpp  -Wl,-rpath,/repositories/halide/build/src  src/libHalide.so.13.0.0 && :
ld: error: /usr/lib/gcc/x86_64-linux-gnu/11/libstdc++.so: undefined reference to pthread_create [--no-allow-shlib-undefined]
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[  2% 35/1480][  0% 0:00:00 + 0:05:09] Generating included_schedule_file.runtime.o
ninja: build stopped: subcommand failed.

```

### CODE REVIEW

**Comment by steven-johnson** [comment]
> while the latter is a dead header not used anywhere

It was used by a handful of tests, but then later removed due to some subtle issues.

**Comment by steven-johnson** [comment]
LGTM pending green buildbots.

**Comment by LebedevRI** [comment]
Do those buildbots report their results to the github commit status? Should i look for their results somewhere else?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -366,6 +366,7 @@ add_library(Halide
             $<TARGET_OBJECTS:Halide_initmod>)
 add_library(Halide::Halide ALIAS Halide)
 
+target_link_libraries(Halide PRIVATE Threads::Threads)
```
Comment (alexreinking) [inline src/CMakeLists.txt:369]: If any of our headers include `<thread>`, then this should be `PUBLIC`. Otherwise, if the only header that includes it doesn't need it, then the include should be dropped and this can remain `PRIVATE`.

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: src/CMakeLists.txt (hunk 1 above)
Comment (LebedevRI) [reply src/CMakeLists.txt:369]: Ah, i see why i couldn't find any includes of the `threadpool.h`, because `halide.h` is an autogenerated amalgamation :/
If i remove that header, building of tests starts failing.

**Comment by LebedevRI** [comment]
@steven-johnson @alexreinking thank you for taking a look!

**Comment by steven-johnson** [comment]
Lots of build failures

File: Makefile [hunk 2]
```diff
@@ -715,7 +715,6 @@ HEADER_FILES = \
   StrictifyFloat.h \
   Substitute.h \
   Target.h \
-  ThreadPool.h \
```
Comment (alexreinking) [inline Makefile:718]: Wait... why did you remove this? I thought that the usage of `<thread>` in here is what was unneeded?

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (LebedevRI) [reply Makefile:718]: Ugh.
The story goes as following: i removed `ThreadPool.h`, noticed that breaks build,
unstaged `ThreadPool.h` removal, then fixed s/PRIVATE/PUBLIC/,
staged changes, checked that it still works for me, and pushed.
The key caveat is that i did not notice that the makefile change got restaged :/

**Review by alexreinking — APPROVED** [review]

**Comment by LebedevRI** [comment]
@alexreinking thank you for the review!
I should have caught all that myself.

**Comment by steven-johnson** [comment]
Still failing, see

https://buildbot.halide-lang.org/master/#/builders/167/builds/162
https://buildbot.halide-lang.org/master/#/builders/167/builds/162/steps/5/logs/stdio

**Comment by alexreinking** [comment]
> Still failing, see

That is a make build and the makefiles were not changed.

**Comment by LebedevRI** [comment]
This looks like something is stale on the buildbot.
The error matches the one from previous build when `ThreadPool.h` was deleted,
but that is clearly not the case in the last commit.

**Comment by steven-johnson** [comment]
trying again via https://github.com/halide/Halide/pull/6259

**Comment by LebedevRI** [comment]
Hurray!

**Comment by LebedevRI** [comment]
@alexreinking i must say, i quite like the UX progress since #4284.
Though, i find that the install footprint is still a bit not right, i'll see if i can provide patches for that.

**Comment by alexreinking** [comment]
> Though, i find that the install footprint is still a bit not right, i'll see if i can provide patches for that.

Is there anything specific you think is wrong? The install rules at this point are highly customizable using the standard GNUInstallDirs hooks

**Comment by LebedevRI** [comment]
Missing "Halide" prefix for autoschedule .so's, docs being installed not into subdirectory, headers not being installed into subdirectory, overly generic names (missing `halide_` prefix perhaps) for autoschedule binaries installed into bin
I think all this should just work out of the box.
